### PR TITLE
Export and import

### PR DESCRIPTION
**Pull request by DGojka**
`Save and export trainings` task: https://trello.com/c/7o3TIhfy/7-save-and-export-trainings
Adding Export and import feature that is used to save (export) trainings to a file or import workouts from a file to the database. The feature is to prevent the loss of training data

### CODE REVIEW

File: app/src/main/java/com/example/stretchy/features/datatransport/DataTransportProperties.kt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package com.example.stretchy.features.datatransport
+
+import android.os.Environment
+
+fun dataTransportFilePath(): String =
+    Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS).absolutePath
+fun dataTransportFileName(): String = "StretchyTrainings.rafalczamp"
```
Comment (PPetka) [inline app/src/main/java/com/example/stretchy/features/datatransport/DataTransportProperties.kt:7]: val, not function

File: app/src/main/java/com/example/stretchy/features/datatransport/Export.kt [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import java.io.File
+
+class Export(val repository: Repository) : ExportSavedData {
+    private val gson = GsonBuilder().setPrettyPrinting().create()
+    private val dataFile = File(dataTransportFilePath(), dataTransportFileName())
+
+    override suspend fun saveDataInFile() {
```
Comment (PPetka) [inline app/src/main/java/com/example/stretchy/features/datatransport/Export.kt:14]: do not launch it here... expose suspenede function and "eat" it later.

File: app/src/main/java/com/example/stretchy/features/datatransport/DataTransportModule.kt [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package com.example.stretchy.features.datatransport
+
+import com.example.stretchy.repository.Repository
+import dagger.Module
+import dagger.Provides
+
+@Module
+class DataTransportModule {
+    @Provides
+    fun provideImport(repository: Repository) = Import(repository)
```
Comment (PPetka) [inline app/src/main/java/com/example/stretchy/features/datatransport/DataTransportModule.kt:10]: `Import` -> `DataImporter`

export the same

File: app/src/main/java/com/example/stretchy/features/datatransport/Export.kt [hunk 4]
```diff
@@ -0,0 +1,29 @@
+package com.example.stretchy.features.datatransport
+
+import com.example.stretchy.repository.Repository
+import com.google.gson.GsonBuilder
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import java.io.File
+
+class Export(val repository: Repository) : ExportSavedData {
```
Comment (PPetka) [inline app/src/main/java/com/example/stretchy/features/datatransport/Export.kt:10]: Interfaces `DataExporter`, `DataImporter`, add `Impl` to name after to implementations

File: app/src/main/java/com/example/stretchy/features/traininglist/di/TrainingListComponent.kt [hunk 5]
```diff
@@ -1,14 +1,15 @@
...
+import com.example.stretchy.features.datatransport.DataTransportModule
 import com.example.stretchy.features.traininglist.ui.TrainingListViewModel
 import dagger.Component
 import javax.inject.Provider
 
 @TrainingListScope
 @Component(
     dependencies = [ActivityComponent::class],
-    modules = [TrainingListModule::class]
+    modules = [TrainingListModule::class,DataTransportModule::class]
```
Comment (PPetka) [inline app/src/main/java/com/example/stretchy/features/traininglist/di/TrainingListComponent.kt:12]: formatting

File: app/src/main/java/com/example/stretchy/features/traininglist/ui/TrainingListViewModel.kt [hunk 6]
```diff
@@ -28,6 +38,19 @@ class TrainingListViewModel(val repository: Repository) : ViewModel() {
         }
     }
 
+    suspend fun import() {
+        viewModelScope.launch {
+            import.addSavedDataToDb()
```
Comment (PPetka) [inline app/src/main/java/com/example/stretchy/features/traininglist/ui/TrainingListViewModel.kt:43]: same here, consume it outside

**Review by PPetka — APPROVED** [review]
- add permissions listener or lambda (to be run on while clicking the buttons). Compose and implement it in activity
- add scope to dagger to make it singleton

**Review by PPetka — CHANGES_REQUESTED** [review]
comment

File: app/src/main/java/com/example/stretchy/activity/MainActivity.kt [hunk 7]
```diff
@@ -1,28 +1,73 @@
...
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         activityComponent = ActivityComponent.create(this)
-
         setContent {
-            Navigation(activityComponent)
+            Navigation(
+                activityComponent,
+                {
+                    askForDataTransportPermissions(WRITE_EXTERNAL_STORAGE)
```
Comment (PPetka) [inline app/src/main/java/com/example/stretchy/activity/MainActivity.kt:33]: `askForDataTransportPermissions` -> `grantPermissions`

File: app/src/main/java/com/example/stretchy/features/datatransport/DataExporter.kt [hunk 8]
```diff
@@ -0,0 +1,5 @@
+package com.example.stretchy.features.datatransport
+
+interface DataExporter {
+    suspend fun saveDataInFile()
+}
```
Comment (PPetka) [inline app/src/main/java/com/example/stretchy/features/datatransport/DataExporter.kt:5]: `saveDataInFile()` -> `exportData`

The file thing is an implementation details, you design interface, the implementation does not need to be file, it could be a cloud solution in future for example

File: app/src/main/java/com/example/stretchy/features/datatransport/DataExporterExporterImpl.kt [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+    override suspend fun saveDataInFile() {
+        val data = getSavedData()
+        dataFile.writeText(data!!)
+    }
+
+    private suspend fun getSavedData(): String? {
+        var data: String? = null
+        CoroutineScope(Dispatchers.Main).launch {
+            data = gson.toJson(repository.getTrainingsWithActivities())
```
Comment (PPetka) [inline app/src/main/java/com/example/stretchy/features/datatransport/DataExporterExporterImpl.kt:22]: Does not consume coroutine here

File: app/src/main/java/com/example/stretchy/features/datatransport/DataImporter.kt [hunk 10]
```diff
@@ -0,0 +1,5 @@
+package com.example.stretchy.features.datatransport
+
+interface DataImporter {
+    suspend fun addSavedDataToDb()
```
Comment (PPetka) [inline app/src/main/java/com/example/stretchy/features/datatransport/DataImporter.kt:4]: same here, `fun importData()`

File: app/src/main/java/com/example/stretchy/features/datatransport/DataImporterImporterImpl.kt [hunk 11]
```diff
@@ -0,0 +1,22 @@
+package com.example.stretchy.features.datatransport
+
+import com.example.stretchy.repository.Repository
+import com.example.stretchy.repository.TrainingWithActivity
+import com.google.gson.Gson
+import com.google.gson.reflect.TypeToken
+import java.io.File
+
+
+class DataImporterImporterImpl(val repository: Repository) : DataImporter {
```
Comment (PPetka) [inline app/src/main/java/com/example/stretchy/features/datatransport/DataImporterImporterImpl.kt:10]: `ImporterImporter`

File: app/src/main/java/com/example/stretchy/features/traininglist/ui/TrainingListViewModel.kt [hunk 12]
```diff
@@ -28,6 +38,19 @@ class TrainingListViewModel(val repository: Repository) : ViewModel() {
         }
     }
 
+    suspend fun import() {
+        viewModelScope.launch {
+            dataImporterImpl.addSavedDataToDb()
+        }.join()
```
Comment (PPetka) [inline app/src/main/java/com/example/stretchy/features/traininglist/ui/TrainingListViewModel.kt:44]: if you do the `launch` here, then the function does not need to be suspended

File: app/src/main/java/com/example/stretchy/features/traininglist/ui/TrainingListViewModel.kt [hunk 13]
```diff
@@ -28,6 +38,19 @@ class TrainingListViewModel(val repository: Repository) : ViewModel() {
...
+        fetchTrainingList()
+    }
+
+    fun export() {
+        viewModelScope.launch {
+            dataExporterImpl.saveDataInFile()
+        }
```
Comment (PPetka) [inline app/src/main/java/com/example/stretchy/features/traininglist/ui/TrainingListViewModel.kt:51]: does it work, hmm I cannot see the collection part

**Review by PPetka — APPROVED** [review]
